### PR TITLE
s_client: Show algorithms & validity period with "Certificate Chain"

### PR DESCRIPTION
**Pull request by MarcT512**
Add certificate validity period (v) and public key & signature algorithms (a) to the s_client "Certificate Chain" output.

Example of connecting to www.google.com:443 showing new (a) & (v) lines:
```
Certificate chain
 0 s:C = US, ST = California, L = Mountain View, O = Google LLC, CN = www.google.com
   i:C = US, O = Google Trust Services, CN = GTS CA 1O1
   a:PKEY: id-ecPublicKey, 256 (bit); sigalg: RSA-SHA256
   v:NotBefore: Dec  3 14:49:26 2019 GMT; NotAfter: Feb 25 14:49:26 2020 GMT
 1 s:C = US, O = Google Trust Services, CN = GTS CA 1O1
   i:OU = GlobalSign Root CA - R2, O = GlobalSign, CN = GlobalSign
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: Jun 15 00:00:42 2017 GMT; NotAfter: Dec 15 00:00:42 2021 GMT
```

### CODE REVIEW

**Comment by richsalz** [comment]
I wonder if it makes sense to add an option that just calls X509 print on each certificate in the chain?  That way you get all information. this one begs the question "why just those fields"?

**Comment by MarcT512** [comment]
Just coming back to this now; apologies for the delay!

> this one begs the question "why just those fields"?

 I was looking to add a bit more detail to the s_client certificate chain output. 

My motivation was an audit of internal servers: looking for any which still had a certificate with a SHA1 signature, and also any certificates due to expire in the next 90 days.

Without this change you have to use -showcerts, save off each certificate and then run it through openssl x509, which is a chore.

> I wonder if it makes sense to add an option that just calls X509 print on each certificate in the chain?

That's certainly a possibility; though produces far more detail than I needed. Happy to do it on another PR if you'd prefer? ie add a "-printcerts" option ?

**Comment by MarcT512** [comment]
NB: I have previously submitted a CLA which was acknowledged (see PR# https://github.com/openssl/openssl/pull/7975 ).

**Comment by t8m** [comment]
@MarcT512 is the commit authorship done with the e-mail address you've specified in the CLA? If not please adjust it. Also there should be no merge commits please.

**Comment by MarcT512** [comment]
Yes, it was submitted against the github email address 34656315+MarcT512@users.noreply.github.com - though this did cause problems previously.

The merge commit was just my attempt at keeping this branch in sync with openssl:master.

**Comment by t8m** [comment]
I've dropped the merge commit however the CLA problem is still there.

It must be something in the CLA check. @mattcaswell @levitte Could you please look?

**Comment by mattcaswell** [comment]
I can see that we had similar CLA problems with your earlier PR #7975. The CLA *is* there in the database and does seem to be correct. In #7975 it seemed that the CLA bot didn't like your email address - but the git commit hooks were fine with it when it came to eventually pushing this.

My guess is that it doesn't like the `+` in the email address. Maybe @levitte can take a look?

**Comment by MarcT512** [comment]
Fixed failing integration tests.

File: apps/s_client.c [hunk 1]
```diff
@@ -3269,6 +3270,17 @@ static void print_stuff(BIO *bio, SSL *s, int full)
                 BIO_printf(bio, "   i:");
                 X509_NAME_print_ex(bio, X509_get_issuer_name(sk_X509_value(sk, i)), 0, get_nameopt());
                 BIO_puts(bio, "\n");
+                public_key = X509_get_pubkey(sk_X509_value(sk, i));
+                if (public_key != NULL) {
+                    BIO_printf(bio, "   a:PKEY: %s, %d (bit); sigalg: %s\n", OBJ_nid2sn(EVP_PKEY_base_id(public_key)), EVP_PKEY_bits(public_key),
```
Comment (beldmit) [inline apps/s_client.c:3275]: This line seems to be too long

Comment (MarcT512) [reply apps/s_client.c]: Agreed - now split over multiple lines.

**Comment by MarcT512** [comment]
Split long line.

**Comment by beldmit** [comment]
Approved implying the CLA trouble is just a mess.

**Comment by MarcT512** [comment]
Is there a way to re-run the integration checks?
Looks like they failed due to 

> fatal: could not create leading directories of 'openssl/openssl': No space left on device

**Comment by beldmit** [comment]
Restarted the failed Travis job

**Comment by beldmit** [comment]
Squashed the last commit and merged.
Many thanks!

**Pull request closed without merging**